File: 040_combination_sum_ii/Solution.go
package main

import (
	"fmt"
	"sort"
)

func combinationSum2(candidates []int, target int) [][]int {
	arrLen := len(candidates)
	var results [][]int
	if arrLen == 0 {
		return results
	}

	var path []int
	sort.Ints(candidates)
	dfs(candidates, 0, arrLen, target, &results, path)
	return results
}

func dfs(candidates []int, begin, arrLen, target int, results *[][]int, path []int) {
	if target == 0 {
		*results = append(*results, append([]int{}, path...))
		return
	}

	for i := begin; i < arrLen; i++ {
		if target < candidates[i] {
			break
		}
		if i > begin && candidates[i] == candidates[i-1] {
			continue
		}

		path = append(path, candidates[i])
		dfs(candidates, i+1, arrLen, target-candidates[i], results, path)
		path = path[:len(path)-1]
	}
}

func main() {
	nums := []int{2, 5, 2, 1, 2}
	target := 5
	results := combinationSum2(nums, target)
	for _, result := range results {
		for _, ele := range result {
			fmt.Printf("%d ", ele)
		}
		fmt.Println()
	}
}

File: 040_combination_sum_ii/Solution.py
class Solution:
    def __init__(self):
        self.path = []
        self.results = []

    def combinationSum2(self, candidates: list, target: int) -> list:
        arr_len = len(candidates)
        if arr_len == 0: return self.results

        candidates.sort()
        self.dfs(candidates, 0, arr_len, target)

        return self.results

    def dfs(self, candidates: list, begin: int, arr_len: int, target: int) -> None:
        if target == 0:
            self.results.append(self.path[:]) # 注意此处 path 需要深拷贝
            return

        for i in range(begin, arr_len):
            if target - candidates[i] < 0: break
            if i > begin and candidates[i] == candidates[i - 1]: continue
            self.path.append(candidates[i])
            self.dfs(candidates, i + 1, arr_len, target - candidates[i])
            self.path.pop()


if __name__ == '__main__':
    solution = Solution()
    results = solution.combinationSum2([2, 5, 2, 1, 2], 5)
    for result in results:
        print(result)

